src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,132 @@
+    public void add(T value, int index) {
+        validateIndexForAdd(index);
+        increaseCapacityIfFull();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
+    }
 
+    @Override
+    public void addAll(MyList<T> list) {
+        while (size + list.size() > elements.length) {
+            grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i)); // Додаємо елементи з переданого списку
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T oldValue = (T) elements[index];
+        System.arraycopy(elements,index + 1, elements, index, size - index - 1);
+        elements[--size] = null;
+        return oldValue;
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public T remove(Object element) {
+        int indexOfElement = indexOf(element);
+        if (indexOfElement == -1) {
+            throw new NoSuchElementException(MESSAGE_NO_ELEMENT);
+        } else {
+            System.arraycopy(elements, indexOfElement + 1, elements, indexOfElement,
+                    size - indexOfElement - 1);
+            elements[--size] = null;
+            return (T) element;
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void validateIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(MESSAGE_OUT_OF_BOUNDS + index);
+        }
+    }
+
+    public void validateIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException(MESSAGE_OUT_OF_BOUNDS + index);
+        }
+    }
+
+    public void increaseCapacityIfFull() {
+        if (size >= elements.length) {
+            grow();
+        }
+    }
+
+    public void grow() {
+        int newCapacity = (int) (elements.length * MULTIPLIER);
+        @SuppressWarnings("unchecked")
+        T[] newArray = (T[]) new Object[newCapacity];
+        System.arraycopy(elements, 0, newArray, 0, size());
+        elements = newArray;
+    }
+
+    public int indexOf(Object o) {
+        if (o == null) {
+            for (int i = 0; i < size; i++) {
+                if (elements[i] == null) {
+                    return i;
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (o.equals(elements[i])) {
+                    return i;
```

Comment:
both clauses have the same body, you need to combine conditions to have only one for loop